File: iOS/Objective C/LevelPlayDemoTests/BannerIntegrationTest.m
#import <XCTest/XCTest.h>
#import "DemoViewController.h"

@interface BannerIntegrationTest : XCTestCase <LPMBannerAdViewDelegate, LPMImpressionDataDelegate>
@property (nonatomic) XCTestExpectation *sdkInitExpectation;
@property (nonatomic) XCTestExpectation *loadExpectation;
@property (nonatomic) XCTestExpectation *impressionExpectation;
@property (nonatomic) LPMBannerAdView *bannerAd;
@end

@implementation BannerIntegrationTest

- (void)testInitLoadAndImpression {
    self.sdkInitExpectation = [self expectationWithDescription:@"Init"];
    self.loadExpectation = [self expectationWithDescription:@"Load"];
    self.impressionExpectation = [self expectationWithDescription:@"Impression"];

    [LevelPlay addImpressionDataDelegate:self];

    LPMInitRequestBuilder *builder = [[LPMInitRequestBuilder alloc] initWithAppKey:kAppKey];
    LPMInitRequest *request = [builder build];

    UIViewController *rootVC = UIApplication.sharedApplication.keyWindow.rootViewController;

    [LevelPlay initWithRequest:request completion:^(LPMConfiguration *config, NSError *error) {
        XCTAssertNil(error, @"Init failed: %@", error);
        [self.sdkInitExpectation fulfill];

        LPMAdSize *bannerSize = [LPMAdSize createAdaptiveAdSize];
        LPMBannerAdViewConfig *bannerConfig = [[[[LPMBannerAdViewConfigBuilder alloc] init] setWithAdSize:bannerSize] build];
        self.bannerAd = [[LPMBannerAdView alloc] initWithAdUnitId:kBannerAdUnitId config:bannerConfig];
        [self.bannerAd setDelegate:self];

        // Add banner to view hierarchy before loading
        dispatch_async(dispatch_get_main_queue(), ^{
            self.bannerAd.translatesAutoresizingMaskIntoConstraints = NO;
            [rootVC.view addSubview:self.bannerAd];
            [NSLayoutConstraint activateConstraints:@[
                [self.bannerAd.centerXAnchor constraintEqualToAnchor:rootVC.view.centerXAnchor],
                [self.bannerAd.bottomAnchor constraintEqualToAnchor:rootVC.view.safeAreaLayoutGuide.bottomAnchor],
                [self.bannerAd.widthAnchor constraintEqualToConstant:320],
                [self.bannerAd.heightAnchor constraintEqualToConstant:50]
            ]];
            [self.bannerAd loadAdWithViewController:rootVC];
        });
    }];

    [self waitForExpectations:@[self.sdkInitExpectation] timeout:10];
    [self waitForExpectations:@[self.loadExpectation] timeout:15];
    [self waitForExpectations:@[self.impressionExpectation] timeout:20];

    // Cleanup
    [self.bannerAd removeFromSuperview];
    [self.bannerAd destroy];
}

#pragma mark - LPMBannerAdViewDelegate

- (void)didLoadAdWithAdInfo:(LPMAdInfo *)adInfo {
    [self.loadExpectation fulfill];
}

- (void)didFailToLoadAdWithAdUnitId:(NSString *)adUnitId error:(NSError *)error {
    XCTFail(@"Load failed: %@", error);
}

- (void)didDisplayAdWithAdInfo:(LPMAdInfo *)adInfo {}
- (void)didFailToDisplayAdWithAdInfo:(LPMAdInfo *)adInfo error:(NSError *)error {}
- (void)didClickAdWithAdInfo:(LPMAdInfo *)adInfo {}
- (void)didExpandAdWithAdInfo:(LPMAdInfo *)adInfo {}
- (void)didCollapseAdWithAdInfo:(LPMAdInfo *)adInfo {}
- (void)didLeaveAppWithAdInfo:(LPMAdInfo *)adInfo {}

#pragma mark - LPMImpressionDataDelegate

- (void)impressionDataDidSucceed:(LPMImpressionData *)impressionData {
    [self.impressionExpectation fulfill];
}

@end

File: iOS/Objective C/LevelPlayDemoTests/BannerIntegrationTest.swift
import Testing
import IronSource
import UIKit

@Test(.timeLimit(.minutes(1)))
@MainActor
func testBannerInitLoadAndImpression() async throws {
    let delegate = BannerTestDelegate()
    let impressionDelegate = BannerImpressionDelegate()

    LevelPlay.add(impressionDelegate as LPMImpressionDataDelegate)

    // Init SDK
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
        let requestBuilder = LPMInitRequestBuilder(appKey: kAppKey)
        let initRequest = requestBuilder.build()
        LevelPlay.initWith(initRequest) { config, error in
            if let error = error {
                continuation.resume(throwing: error)
            } else {
                continuation.resume()
            }
        }
    }

    // Create banner ad using adaptive size (matching the ObjC demo)
    let bannerSize = LPMAdSize.createAdaptive()!
    let config = LPMBannerAdViewConfigBuilder().set(adSize: bannerSize).build()
    let bannerAd = LPMBannerAdView(adUnitId: kBannerAdUnitId, config: config)
    bannerAd.setDelegate(delegate)

    // Add banner to view hierarchy before loading
    let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene
    let rootVC = windowScene?.windows.first(where: { $0.isKeyWindow })?.rootViewController

    bannerAd.translatesAutoresizingMaskIntoConstraints = false
    rootVC!.view.addSubview(bannerAd)
    NSLayoutConstraint.activate([
        bannerAd.centerXAnchor.constraint(equalTo: rootVC!.view.centerXAnchor),
        bannerAd.bottomAnchor.constraint(equalTo: rootVC!.view.safeAreaLayoutGuide.bottomAnchor),
        bannerAd.widthAnchor.constraint(equalToConstant: CGFloat(bannerSize.width)),
        bannerAd.heightAnchor.constraint(equalToConstant: CGFloat(bannerSize.height))
    ])

    // Load ad (banner auto-displays on load)
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
        delegate.onLoad = { continuation.resume() }
        delegate.onLoadFail = { error in continuation.resume(throwing: error) }
        bannerAd.loadAd(with: rootVC!)
    }

    // Wait for impression
    if !impressionDelegate.received {
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            if impressionDelegate.received {
                continuation.resume()
            } else {
                impressionDelegate.onImpression = { continuation.resume() }
            }
        }
    }

    // Cleanup
    bannerAd.removeFromSuperview()
    bannerAd.destroy()
}

private class BannerTestDelegate: NSObject, LPMBannerAdViewDelegate {
    var onLoad: (() -> Void)?
    var onLoadFail: ((Error) -> Void)?

    func didLoadAd(with adInfo: LPMAdInfo) { onLoad?() }
    func didFailToLoadAd(withAdUnitId adUnitId: String, error: Error) { onLoadFail?(error) }
    func didDisplayAd(with adInfo: LPMAdInfo) {}
    func didFailToDisplayAd(with adInfo: LPMAdInfo, error: Error) {}
    func didClickAd(with adInfo: LPMAdInfo) {}
    func didExpandAd(with adInfo: LPMAdInfo) {}
    func didCollapseAd(with adInfo: LPMAdInfo) {}
    func didLeaveApp(with adInfo: LPMAdInfo) {}
}

private class BannerImpressionDelegate: NSObject, LPMImpressionDataDelegate {
    var onImpression: (() -> Void)?
    private(set) var received = false

    func impressionDataDidSucceed(_ impressionData: LPMImpressionData) {
        received = true
        onImpression?()
    }
}
